HighAvailabilityModule.Server.InMemory/InMemoryMembershipServer.cs

```diff
@@ -22,53 +24,78 @@ public InMemoryMembershipServer(TimeSpan timeout)
 
         public void RemoveCurrent()
         {
-            lock (this.heartbeatLock)
+            lock (heartbeatLock)
             {
-                this.Current = HeartBeatEntry.Empty;
+                List<string> key = new List<string>(this.CurrentTable.Keys);
+                for (int i = 0; i < key.Count; i++)
+                {
+                    this.CurrentTable[key[i]] = HeartBeatEntry.Empty;
+                }
             }
         }
 
-        public Task HeartBeatAsync(string uuid, HeartBeatEntry lastSeenEntry) => this.HeartBeatAsync(uuid, lastSeenEntry, DateTime.UtcNow);
+        public Task HeartBeatAsync(HeartBeatEntryDTO entryDTO) => this.HeartBeatAsync(entryDTO, DateTime.UtcNow);
 
-        public async Task HeartBeatAsync(string uuid, HeartBeatEntry lastSeenEntry, DateTime now)
+        public async Task HeartBeatAsync(HeartBeatEntryDTO entryDTO, DateTime now)
         {
             bool ValidInput()
             {
-                return this.Current == null
-                       || (this.HeartbeatInvalid(now) && lastSeenEntry != null && lastSeenEntry.IsEmpty)
-                       || (lastSeenEntry != null && this.Current.Uuid == lastSeenEntry.Uuid && this.Current.TimeStamp == lastSeenEntry.TimeStamp && this.Current.Uuid == uuid);
+                if (this.CurrentTable.ContainsKey(entryDTO.Utype))
+                {
+                    return this.CurrentTable[entryDTO.Utype] == null
```

Comment:
This logic has grown fairly complicated. The best practice here is to extract it to a separate method.